### ISSUE TITLE
In-memory CRUD Operations for Generated Royale Interface

### ISSUE THREAD

**Issue by JoelProminic** [Moonshine-IDE/Moonshine-IDE#1040]
This is a sub-task of #675.  We discussed this in the meeting today, but this is complex enough that I wanted to pull it into a separate issue.

We would like to have an in-memory object model and CRUD interface as part of the generated Royale application.  The main goals are:
- Allow the user (in the short term, developers) to run more practical tests against the generated forms
- Make it more comfortable for the user to customize an maintain the generated application later.
- Make a placeholder for the CRUD agents (#978) until this is ready.  Ideally, this code should be written to make it easier to interface with the agents later

@rat-moonshine implemented something similar to this for Form Builder in #704.

The ValueObject properties can be built from the fields (DominoInputText) found in the Domino Visual Editor intermediate XML.  You should mainly need the Name, Type, and Multivalue properties for this.

One feature that is missing currently in Domino Visual Editor is view configuration.  For an initial implementation, just include the first X fields from the form in the View.  

Similarly, we don't define keys, which will be needed for the CRUD agents.  You can ignore this for now in the view.

**Comment by piotrzarzycki21** [comment]
Next step would be generate usage of this VO in each view.

**Comment by piotrzarzycki21** [comment]
@JoelProminic I have started to take care a little bit of data type of fields in VO, cause it seems to me it is started to be important. I have following code:

```
<par def="1013" paragraph="true" dominotype="paragraph" class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop">
              <field useappletinbrowser="false" allowtabout="false" defaultfocus="false" protected="false" sign="false" storelocally="false" value="Disable" type="keyword" kind="editable" computeaftervalidation="false" allowmultivalues="true" width="100pt" height="30pt" bgcolor="#ffffff" name="Disable" borderstyle="inset" listdisplayseparator="comma" listinputseparators="comma" lookupaddressonrefresh="false" lookupeachchar="false" showdelimiters="true">
                <keywords ui="checkbox" recalconchange="false" recalcchoices="false" columns="1">
                  <textlist>
                    <text>True | true</text>
                  </textlist>
                </keywords>
              </field>
            </par>
```

In the result I will have following field in VO:
`public var Disable:Boolean = "Disable";`

Value is incorrect it should be true/false - Should I take that value from some different property above or should I leave default false ?

**Comment by piotrzarzycki21** [comment]
I have pushed for now default value "false" for Boolean type - Let me know how it should be.

**Comment by piotrzarzycki21** [comment]
I have made another step to get this working. Generate VO now is being used in MXML - I believe next step would be connect with server sight.

**Comment by JoelProminic** [comment]
@piotrzarzycki21, I mentioned the CRUD interface in the description, but from our meeting today it seems that this was not clear.  I'm envisioning an interface similar to what we generated for FormBuilder in #704. The components are:
- A menu to select a form.  You already have this with the sidebar
- A view to list the available documents/objects for that form.  We aren't configuring this currently in Domino Visual Editor, but I'll create a separate issue to discuss how we can generate something similar for this.  
- A collection of value objects for the view
- A form - this has been our primary focus so far.  This should have buttons to save or navigate back to the view
- A value object for the form
- Actions to create, edit or delete documents/objects


What configuration information is missing to generate this full interface.  The only thing I'm seeing is the view configuration, including:
- field subset
- Ordering
- sorting/catagorization

A placeholder solution would be to make this a single-column datagrid or list with a generic label (i.e. "Object1", "Object2").  This wouldn't be desirable for a real application, but it would let us continue with the CRUD interface until we have a better solution.

You may be able to reuse the view and action logic from the templates in #704.  Talk to @rat-moonshine if you need clarification for this.

We also want to be able to generate the Royale application to read from generated Domino agents instead.

**Comment by JoelProminic** [comment]
As discussed in the meeting, I recommend you use a placeholder datagrid for now that will just serve as a list of available objects.

For example, the interface might look like this:
![image](https://user-images.githubusercontent.com/24532731/176792932-6d3a20f0-cd70-4072-b0b7-a138a6614ba6.png)

Then when the user clicks edits an (double-click), it opens the form you generated, bound to the selected object.  New/Create would open a new object.
![image](https://user-images.githubusercontent.com/24532731/176793104-c84fce19-2854-4d57-adbb-be49aa0f2aaa.png)

The values in the Authors datagrid doesn't matter for now.  This can be a constant string (like what I did), an object reference, or just the first property from the form/value object.  This will allow you to continue working on the CRUD actions until we have a more permanent solution for defining the view (datagrid) from #1057.

**Comment by piotrzarzycki21** [comment]
@JoelProminic I have finished adding to generated Royale application "Edit" option. It should be functional and ready for review. I will try to handle option "Add" today as well.

**Comment by piotrzarzycki21** [comment]
@JoelProminic I have fixed problems with editing in more sophisticated views and displaying more columns in DataGrid.

**Comment by piotrzarzycki21** [comment]
@JoelProminic I have pushed ability to remove items from list - it's ready for tests. What should be done next actually ?

**Comment by JoelProminic** [comment]
When I did a fresh test for the PerformanceStats_BingTest database, I got this error on compilation:

: /Users/joelanderson/Documents/Flex/temp_workspace/PerformanceStats_BingTest_Royale/src/PerformanceStats_BingTest_Royale.mxml(72): col: 5 Error: This tag is unexpected. It will be ignored.
:     <view:ConfigValue xmlns:view="view.*"/>
:     ^

I figured out that this was because of the space in the class name.  You should strip out spaces and other invalid special characters when generating the class names.  I also found references in the class that were broken.  For example:

```
	[Bindable] private var config valueVO:Config ValueVO;
	[Bindable] private var config valueVOItems:Array = [new Config ValueVO()];
```

I tried to rename this from the class, but I found that the case-sensitive checkbox in the Moonshine Find/Replace did not work - I may report this separately.

I renamed `nsfs/nsf-moonshine/odp/Forms/Config Value.form` and `visualeditor-src/main/webapp/Config Value.xml` in my test converted database to remove the space for this test.  There is a minor bug with this rename that we can fix separately (it is doing a replaceAll on ".form" as a regular expression):

    : Unable to access file:/Users/-user-/Documents/Flex/temp_workspace/Pe.xmlanceStats_BingTest/visualeditor-src/main/webapp/Config Value.form
    : Unable to access file:/Users/-user-/Documents/Flex/temp_workspace/Pe.xmlanceStats_BingTest/visualeditor-src/main/webapp/Config Value.xml

With this change, I was able to build and run the project.  I confirmed that the CRUD operations were working for the ConfigValue form.

Some issues:
- Updates to "richtext" fields are not saved during Add or Update.  For an example, see Differences on the first tab of HostConfigSnapshot.
- I'm still seeing issues with the line breaks (#1033) in the HostConfigSnapshot form.  I see that my generated form has "$hostname" and the HostnameOutput in separate Paragraphs here:

![image](https://user-images.githubusercontent.com/24532731/180879958-fa8a9217-b385-4996-b857-8df8d371208d.png)

But they are displayed on the same line in Royale:

![image](https://user-images.githubusercontent.com/24532731/180879938-1d98c652-f4b0-44a1-845a-1b99b97245e8.png)

Are you still waiting on me for a fresh export of the PerformanceStats_BingTest with these newline fixes?  I tested on a fresh conversion on my side.

**Comment by JoelProminic** [comment]
Some bigger changes that are still pending before we can start on integration with the agents:
- View Builder (#1057) - I need to revisit this soon
- Default Value.  Currently this is populated from the "Value" property, but this is not correct for Domino Visual Editor.  The actual default value is based on a Domino formula (Field Info > Client Objects > defaultvalue > Formula).  Don't worry about parsing this formula for now - it will be difficult to handle cases beyond no value and a string or number literal.   The text default value will cause problems for at least the type=number fields as we move to the agents.  

![image](https://user-images.githubusercontent.com/24532731/180881206-da5384d2-f7e5-460e-9807-f3f6d7cdc1da.png)

**Comment by piotrzarzycki21** [comment]
> I renamed `nsfs/nsf-moonshine/odp/Forms/Config Value.form` and `visualeditor-src/main/webapp/Config Value.xml` in my test converted database to remove the space for this test. There is a minor bug with this rename that we can fix separately (it is doing a replaceAll on ".form" as a regular expression):

I'm not sure who should fix that issue? Shouldn't I get from converter project which doesn't contains such things ? I believe that should be the case. We do not allow create files with spaces in Moonshine, so that's the rule where converter should adjust itself in my opinion.

**Comment by piotrzarzycki21** [comment]
> Are you still waiting on me for a fresh export of the PerformanceStats_BingTest with these newline fixes? I tested on a fresh conversion on my side.

I think I do. I haven't have fresh copy of that project for a long time.

**Comment by piotrzarzycki21** [comment]
> * Default Value.  Currently this is populated from the "Value" property, but this is not correct for Domino Visual Editor.  The actual default value is based on a Domino formula (Field Info > Client Objects > defaultvalue > Formula).  Don't worry about parsing this formula for now - it will be difficult to handle cases beyond no value and a string or number literal.   The text default value will cause problems for at least the type=number fields as we move to the agents.

@JoelProminic please raise separate issue for this - I would like to discuss it widely.

**Comment by piotrzarzycki21** [comment]
I think we should close this issue and continue in new one.

**Comment by piotrzarzycki21** [comment]
> When I did a fresh test for the PerformanceStats_BingTest database, I got this error on compilation:
> 
> : /Users/joelanderson/Documents/Flex/temp_workspace/PerformanceStats_BingTest_Royale/src/PerformanceStats_BingTest_Royale.mxml(72): col: 5 Error: This tag is unexpected. It will be ignored. : <view:ConfigValue xmlns:view="view.*"/> : ^
> 
> I figured out that this was because of the space in the class name. You should strip out spaces and other invalid special characters when generating the class names. I also found references in the class that were broken. For example:
> 
> ```
> 	[Bindable] private var config valueVO:Config ValueVO;
> 	[Bindable] private var config valueVOItems:Array = [new Config ValueVO()];
> ```
> 
> I tried to rename this from the class, but I found that the case-sensitive checkbox in the Moonshine Find/Replace did not work - I may report this separately.
> 
> I renamed `nsfs/nsf-moonshine/odp/Forms/Config Value.form` and `visualeditor-src/main/webapp/Config Value.xml` in my test converted database to remove the space for this test. There is a minor bug with this rename that we can fix separately (it is doing a replaceAll on ".form" as a regular expression):
> 
> ```
> : Unable to access file:/Users/-user-/Documents/Flex/temp_workspace/Pe.xmlanceStats_BingTest/visualeditor-src/main/webapp/Config Value.form
> : Unable to access file:/Users/-user-/Documents/Flex/temp_workspace/Pe.xmlanceStats_BingTest/visualeditor-src/main/webapp/Config Value.xml
> ```
> 
> With this change, I was able to build and run the project. I confirmed that the CRUD operations were working for the ConfigValue form.
> 
> Some issues:
> 
> * Updates to "richtext" fields are not saved during Add or Update.  For an example, see Differences on the first tab of HostConfigSnapshot.
> * I'm still seeing issues with the line breaks ([Inconsistent Newline Behavior in Converted Domino Database #1033](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1033)) in the HostConfigSnapshot form.  I see that my generated form has "$hostname" and the HostnameOutput in separate Paragraphs here:
> 
> ![image](https://user-images.githubusercontent.com/24532731/180879958-fa8a9217-b385-4996-b857-8df8d371208d.png)
> 
> But they are displayed on the same line in Royale:
> 
> ![image](https://user-images.githubusercontent.com/24532731/180879938-1d98c652-f4b0-44a1-845a-1b99b97245e8.png)
> 
> Are you still waiting on me for a fresh export of the PerformanceStats_BingTest with these newline fixes? I tested on a fresh conversion on my side.

I took a deeper look into Paragraph and why they are not in new line. Mentioned by you hostname code looks like that:

```
<Column>
                <Div class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" height="300px" percentWidth="100">
                  <Paragraph alignPardef="" class="flexHorizontalLayout flexHorizontalLayoutLeft flexHorizontalLayoutTop" defid="11" isNewLine="true" wrap="false">
                    <run hide="">
                      <font name="monospace"/>
                      <Label color="system" fontName="monospace" hide="" isBreak="false" isNewLine="false" isUrlLink="false" style="normal">$ hostname</Label>
                    </run>
                  </Paragraph>
                  <Paragraph isNewLine="true">
                    <run hide="">
                      <font name="monospace"/>
                      <Field fontName="monospace" height="30" hidewhen="" htmlClass="" htmlId="" htmlOther="" htmlStyle="" htmlTitle="" id="" kind="editable" listdisplayseparator="" listinputseparators="" maxlength="" name="HostnameOutput" required="false" type="text" value="HostnameOutput" width="106"/>
                    </run>
                    <run hide="">
                      <font name="monospace"/>
                    </run>
                  </Paragraph>
                </Div>
</Column>
```

These paragraphs are part of DominoTable which contains Column and each column has internal container Div - which has horizontal layout. I think we should change that and we will have good behavior. We could agreed on having DominoTable with vertical layout inside rows/columns.

**Comment by piotrzarzycki21** [comment]
I have moved discussion about new line behaviour to #1033 . I will close this issue. @JoelProminic please take a look into my above comments. I will raise also issue for default value.
